src/object/cli_obj.c

```diff
@@ -1509,6 +1521,415 @@ shard_task_reset_target(tse_task_t *shard_task, void *arg)
 	return 0;
 }
 
+/* EC struct used to save state during encoding and to drive resource recovery.
+ */
+struct ec_params {
+	daos_iod_t		*iods;	/* Replaces iod array in update.
+					 * NULL except head of list
+					 */
+	daos_sg_list_t		*sgls;	/* Replaces sgl array in update.
+					 * NULL except head
+					 */
+	unsigned int		nr;	/* number of records in iods and sgls
+					 * (same as update_t)
+					 */
+	daos_iod_t		niod;	/* replacement IOD for an input IOD
+					 * that includes full stripe.
+					 */
+	daos_sg_list_t		nsgl;	/* replacement SGL for an input IOD that
+					 * includes full stripe.
+					 */
+	struct obj_ec_parity	p_segs;	/* Structure containing array of
+					 * pointers to parity extents.
+					 */
+	struct ec_params        *next;	/* Pointer to next entry in list. */
+};
+
+/* Determines weather a given IOD contains a recx that is at least a full
+ * stripe's worth of data.
+ */
+static bool
+ec_has_full_stripe(daos_iod_t *iod, struct daos_oclass_attr *oca,
+		   uint64_t *tgt_set)
+{
+	unsigned int	ss = oca->u.ec.e_k * oca->u.ec.e_len;
+	unsigned int	i;
+
+	for (i = 0; i < iod->iod_nr; i++) {
+		if (iod->iod_type == DAOS_IOD_ARRAY) {
+			int start = iod->iod_recxs[i].rx_idx * iod->iod_size;
+			int length = iod->iod_recxs[i].rx_nr * iod->iod_size;
+
+			if (length < ss) {
+				continue;
+			} else if (length - (start % ss) >= ss) {
+				*tgt_set = ~0UL;
```

Comment:
hmm, the bit in tgt_set stands for which target the I/O will touch?  then I thought you only need set bits of (e_k+e_p)? instead of the whole bits?  or I miss sth?

Reply:
This is a bit convoluted.  There are comments there to explain.  The tgt_set is initialized to zero. Non-EC updates always go to all targets, so all zeros as init value means this will happen. 

For EC updates, non-zero means full stripe, and the code in dc_obj_update sets the the tgt_set back to zero.  It is still zero if not a full stripe. Then the ec_init_tgt_set has to set correct bits.  For full stripe we can treat the target selection the same as for replica objects.  Hence, I don't need to set the correct bits for full stripe, I just set it to non-zero in the ec_update call, and use this to decide whether we need to check for specific targets. Since we don't (since it's full stripe), we set the tgt_set back to zero.  This makes the tgt selection act the same for replica objects and full-stripe updates.

Reply:
ah, so except the tgt bitmap, you also use it as a "flag". Hmm, you can set bits for both non-EC update and full stripe EC updates, which might a bit cleaner? Anyway, it is not a big deal.
Hmm, probably we should separate forward targets building for ec and non-ec update,

Reply:
I was mostly interested in avoiding the pass over the recxs to determine targets if I already knew it was going to all of them.  I realized after I had it coded and tested that it was confusing, so I added the comments to explain.  I agree that it should probably be re-implemented at some point.